File: codes/visual_basic_dotnet/array_map.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function JsonStringify(ByVal Anything As Object, Optional ByVal Pretty As Boolean = False, Optional ByVal Indent As String = "    ") As String
        Dim IndentLevel As Integer = 0
        Dim JsonStringifyInner As Func(Of Object, String, String) = Function(ByVal AnythingInner As Object, ByVal IndentInner As String)
            If (AnythingInner Is Nothing) Then Return "null"
            If (TypeOf AnythingInner Is String) Then Return """" & AnythingInner & """"
            If (IsNumeric(AnythingInner) = True) Then Return CStr(AnythingInner).Replace(",", ".")
            If (TypeOf AnythingInner Is Boolean) Then Return CStr(AnythingInner)
            If (TypeOf AnythingInner Is List(Of Object)) Then
                If (AnythingInner.Count = 0) Then Return "[]"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "[" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "[")
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In AnythingInner
                    Result &= JsonStringifyInner(ArrayItem, IndentInner)
                    If ((ArrayItemIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ArrayItemIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "]", "]")
                Return Result
            End If
            If (TypeOf AnythingInner Is Dictionary(Of String, Object)) Then
                If (AnythingInner.Count = 0) Then Return "{}"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "{" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "{")
                Dim ObjectIterationIndex As Integer = 0
                For Each ObjectEntry As KeyValuePair(Of String, Object) In AnythingInner
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    Result &= """" & ObjectKey & """: " & JsonStringifyInner(ObjectValue, IndentInner)
                    If ((ObjectIterationIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ObjectIterationIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "}", "}")
                Return Result
            End If
            Return "null"
        End Function
        Return JsonStringifyInner(Anything, Indent)
    End Function

    Function SpreadObject(ByVal ParamArray Arguments() As Object) As Dictionary(Of String, Object)
        Dim NewObject As New Dictionary(Of String, Object)()
        Dim ArgumentIndex As Integer = 0
        For Each Argument As Object In Arguments
            If (TypeOf Argument Is Dictionary(Of String, Object)) Then
                For Each ObjectEntry As KeyValuePair(Of String, Object) In Argument
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    NewObject(CStr(ObjectKey)) = ObjectValue
                Next
                Continue For
            End If
            If (TypeOf Argument Is List(Of Object)) Then
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In Argument
                    NewObject(CStr(ArrayItemIndex)) = ArrayItem
                    ArrayItemIndex += 1
                Next
                Continue For
            End If
            ArgumentIndex += 1
        Next
        Return NewObject
    End Function

    Function ArrayMapV1(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Object), ByVal AnArray As List(Of Object)) As List(Of Object)
        ' JavaScript-like Array.map() function
        Dim NewArray As New List(Of Object)()
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            Dim NewArrayItem As Object = CallbackFunction(ArrayItem, ArrayItemIndex, AnArray)
            NewArray.Add(NewArrayItem)
            ArrayItemIndex += 1
        Next
        Return NewArray
    End Function

    Function ArrayMapV2(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Object), ByVal AnArray As List(Of Object)) As List(Of Object)
        ' JavaScript-like Array.map() function
        Dim NewArray As New List(Of Object)()
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            NewArray.Add(CallbackFunction(ArrayItem, ArrayItemIndex, AnArray))
            ArrayItemIndex += 1
        Next
        Return NewArray
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.map() in Visual Basic (.NET) List")

        Dim Numbers As List(Of Object) = New List(Of Object) From {12, 34, 27, 23, 65, 93, 36, 87, 4, 254}
        Console.WriteLine($"Numbers: {JsonStringify(Numbers)}")

        Dim NumbersLabeled As New List(Of Object)()

        Console.WriteLine("' using JavaScript-like Array.map() function ""ArrayMapV1""")

        NumbersLabeled = ArrayMapV1(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) New Dictionary(Of String, Object) From {{CStr(Number), If(((Number Mod 2) = 0), "even", "odd")}}, Numbers)
        Console.WriteLine($"labeled numbers: {JsonStringify(NumbersLabeled, Pretty:=True)}")
        ' labeled numbers: [
        '     {
        '         "12": "even"
        '     },
        '     {
        '         "34": "even"
        '     },
        '     {
        '         "27": "odd"
        '     },
        '     {
        '         "23": "odd"
        '     },
        '     {
        '         "65": "odd"
        '     },
        '     {
        '         "93": "odd"
        '     },
        '     {
        '         "36": "even"
        '     },
        '     {
        '         "87": "odd"
        '     },
        '     {
        '         "4": "even"
        '     },
        '     {
        '         "254": "even"
        '     }
        ' ]

        Console.WriteLine("' using JavaScript-like Array.map() function ""ArrayMapV2""")

        NumbersLabeled = ArrayMapV2(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) New Dictionary(Of String, Object) From {{CStr(Number), If(((Number Mod 2) = 0), "even", "odd")}}, Numbers)
        Console.WriteLine($"labeled numbers: {JsonStringify(NumbersLabeled, Pretty:=True)}")
        ' labeled numbers: [
        '     {
        '         "12": "even"
        '     },
        '     {
        '         "34": "even"
        '     },
        '     {
        '         "27": "odd"
        '     },
        '     {
        '         "23": "odd"
        '     },
        '     {
        '         "65": "odd"
        '     },
        '     {
        '         "93": "odd"
        '     },
        '     {
        '         "36": "even"
        '     },
        '     {
        '         "87": "odd"
        '     },
        '     {
        '         "4": "even"
        '     },
        '     {
        '         "254": "even"
        '     }
        ' ]

        Console.WriteLine("' using Visual Basic (.NET) Array.map() built-in method ""Array.Select().Cast(Of Object)().ToList()""")

        NumbersLabeled = Numbers.Select(Function(ByVal Number As Integer) New Dictionary(Of String, Object) From {{CStr(Number), If(((Number Mod 2) = 0), "even", "odd")}}).Cast(Of Object)().ToList()
        Console.WriteLine($"labeled numbers: {JsonStringify(NumbersLabeled, Pretty:=True)}")
        ' labeled numbers: [
        '     {
        '         "12": "even"
        '     },
        '     {
        '         "34": "even"
        '     },
        '     {
        '         "27": "odd"
        '     },
        '     {
        '         "23": "odd"
        '     },
        '     {
        '         "65": "odd"
        '     },
        '     {
        '         "93": "odd"
        '     },
        '     {
        '         "36": "even"
        '     },
        '     {
        '         "87": "odd"
        '     },
        '     {
        '         "4": "even"
        '     },
        '     {
        '         "254": "even"
        '     }
        ' ]

        Console.WriteLine("' using Visual Basic (.NET) Array.map() built-in method ""Array.Select(DirectCast()).ToList()""")

        NumbersLabeled = Numbers.Select(Function(ByVal Number As Integer) DirectCast(New Dictionary(Of String, Object) From {{CStr(Number), If(((Number Mod 2) = 0), "even", "odd")}}, Object)).ToList()
        Console.WriteLine($"labeled numbers: {JsonStringify(NumbersLabeled, Pretty:=True)}")
        ' labeled numbers: [
        '     {
        '         "12": "even"
        '     },
        '     {
        '         "34": "even"
        '     },
        '     {
        '         "27": "odd"
        '     },
        '     {
        '         "23": "odd"
        '     },
        '     {
        '         "65": "odd"
        '     },
        '     {
        '         "93": "odd"
        '     },
        '     {
        '         "36": "even"
        '     },
        '     {
        '         "87": "odd"
        '     },
        '     {
        '         "4": "even"
        '     },
        '     {
        '         "254": "even"
        '     }
        ' ]

        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.map() in Visual Basic (.NET) List of Dictionaries")

        Dim Products As List(Of Object) = New List(Of Object) From {
            New Dictionary(Of String, Object) From {
                {"code", "pasta"},
                {"price", 321}
            },
            New Dictionary(Of String, Object) From {
                {"code", "bubble_gum"},
                {"price", 233}
            },
            New Dictionary(Of String, Object) From {
                {"code", "potato_chips"},
                {"price", 5}
            },
            New Dictionary(Of String, Object) From {
                {"code", "towel"},
                {"price", 499}
            }
        }
        Console.WriteLine($"Products: {JsonStringify(Products, Pretty:=True)}")

        Dim ProductsLabeled As New List(Of Object)()

        Console.WriteLine("' using JavaScript-like Array.map() function ""ArrayMapV1""")

        ProductsLabeled = ArrayMapV1(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) SpreadObject(Product, New Dictionary(Of String, Object) From {{"label", If((Product("price") > 100), "expensive", "cheap")}}), Products)
        Console.WriteLine($"labeled products: {JsonStringify(ProductsLabeled, Pretty:=True)}")
        ' labeled products: [
        '     {
        '         "code": "pasta",
        '         "price": 321,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "bubble_gum",
        '         "price": 233,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "potato_chips",
        '         "price": 5,
        '         "label": "cheap"
        '     },
        '     {
        '         "code": "towel",
        '         "price": 499,
        '         "label": "expensive"
        '     }
        ' ]

        Console.WriteLine("' using JavaScript-like Array.map() function ""ArrayMapV2""")

        ProductsLabeled = ArrayMapV2(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) SpreadObject(Product, New Dictionary(Of String, Object) From {{"label", If((Product("price") > 100), "expensive", "cheap")}}), Products)
        Console.WriteLine($"labeled products: {JsonStringify(ProductsLabeled, Pretty:=True)}")
        ' labeled products: [
        '     {
        '         "code": "pasta",
        '         "price": 321,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "bubble_gum",
        '         "price": 233,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "potato_chips",
        '         "price": 5,
        '         "label": "cheap"
        '     },
        '     {
        '         "code": "towel",
        '         "price": 499,
        '         "label": "expensive"
        '     }
        ' ]

        Console.WriteLine("' using Visual Basic (.NET) Array.map() built-in method ""Array.Select().Cast(Of Object)().ToList()""")

        ProductsLabeled = Products.Select(Function(ByVal Product As Object) SpreadObject(Product, New Dictionary(Of String, Object) From {{"label", If((Product("price") > 100), "expensive", "cheap")}})).Cast(Of Object)().ToList()
        Console.WriteLine($"labeled products: {JsonStringify(ProductsLabeled, Pretty:=True)}")
        ' labeled products: [
        '     {
        '         "code": "pasta",
        '         "price": 321,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "bubble_gum",
        '         "price": 233,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "potato_chips",
        '         "price": 5,
        '         "label": "cheap"
        '     },
        '     {
        '         "code": "towel",
        '         "price": 499,
        '         "label": "expensive"
        '     }
        ' ]

        Console.WriteLine("' using Visual Basic (.NET) Array.map() built-in method ""Array.Select(DirectCast()).ToList()""")

        ProductsLabeled = Products.Select(Function(ByVal Product As Object) DirectCast(SpreadObject(Product, New Dictionary(Of String, Object) From {{"label", If((Product("price") > 100), "expensive", "cheap")}}), Object)).ToList()
        Console.WriteLine($"labeled products: {JsonStringify(ProductsLabeled, Pretty:=True)}")
        ' labeled products: [
        '     {
        '         "code": "pasta",
        '         "price": 321,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "bubble_gum",
        '         "price": 233,
        '         "label": "expensive"
        '     },
        '     {
        '         "code": "potato_chips",
        '         "price": 5,
        '         "label": "cheap"
        '     },
        '     {
        '         "code": "towel",
        '         "price": 499,
        '         "label": "expensive"
        '     }
        ' ]
    End Sub
End Module

File: codes/gnu_octave/array_map.m
function stringrepeatresult = stringrepeat(astring, count)
    result = "";
    for i = (1:1:count) % (start:step:stop)
        result = cstrcat(result, astring);
    end
    stringrepeatresult = result;
end

function arrayreduceresult = arrayreduce(callbackfunction, anarray, initialvalue)
    % JavaScript-like Array.reduce() function
    currentresult = initialvalue;
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        currentresult = callbackfunction(currentresult, arrayitem, arrayitemindex, anarray);
    end
    arrayreduceresult = currentresult;
end

function optionalchainingresult = optionalchaining(anything, varargin)
    % JavaScript-like Optional Chaining Operator (?.) function
    % varargin == objectpropertiesarray
    if ((~isstruct(anything) && ~iscell(anything)) || isempty(varargin))
        optionalchainingresult = anything;
        return;
    end
    function arrayreducecallbackresult = arrayreducecallback(currentresult, currentitem, varargin)
        % varargin == (arrayitemindex, anarray)
        if ((isempty(currentresult) && isstruct(anything) && (ischar(currentitem) || ischar(currentitem))) && isfield(anything, currentitem))
            arrayreducecallbackresult = anything.(currentitem);
            return;
        end
        if (isempty(currentresult) && iscell(anything) && isnumeric(currentitem) && (currentitem >= 1) && (numel(anything) > currentitem))
            arrayreducecallbackresult = anything{currentitem};
            return;
        end
        if ((isstruct(currentresult) && (ischar(currentitem) || ischar(currentitem))) && isfield(currentresult, currentitem))
            arrayreducecallbackresult = currentresult.(currentitem);
            return;
        end
        if (iscell(currentresult) && isnumeric(currentitem) && (currentitem >= 1) && (numel(currentresult) > currentitem))
            arrayreducecallbackresult = currentresult{currentitem};
            return;
        end
        arrayreducecallbackresult = {};
    end
    optionalchainingresult = arrayreduce(@arrayreducecallback, varargin, {});
end

function ternaryresult = ternary(truecondition, callbackfunctionifconditiontrue, callbackfunctionifconditionfalse)
    if (islogical(truecondition) && (truecondition == true))
        ternaryresult = callbackfunctionifconditiontrue();
        return;
    end
    ternaryresult = callbackfunctionifconditionfalse();
end

function jsonstringifyresult = jsonstringify(anything, varargin)
    indentlevel = 0;
    function jsonstringifyinnerresult = jsonstringifyinner(anythinginner, prettyinner, indentinner)
        if isempty(anythinginner)
            jsonstringifyinnerresult = "null";
            return;
        end
        if (ischar(anythinginner) || ischar(anythinginner))
            jsonstringifyinnerresult = cstrcat("""", anythinginner, """");
            return;
        end
        if isnumeric(anythinginner)
            jsonstringifyinnerresult = num2str(anythinginner);
            return;
        end
        if (islogical(anythinginner) && (anythinginner == true))
            jsonstringifyinnerresult = "true";
            return;
        end
        if (islogical(anythinginner) && (anythinginner == false))
            jsonstringifyinnerresult = "false";
            return;
        end
        if iscell(anythinginner)
            if isempty(anythinginner)
                jsonstringifyinnerresult = "[]";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("[", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("["));
            for arrayitemindex = (1:1:numel(anythinginner)) % (start:step:stop)
                arrayitem = anythinginner{arrayitemindex};
                result = cstrcat(result, jsonstringifyinner(arrayitem, prettyinner, indentinner));
                if (arrayitemindex ~= numel(anythinginner))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "]")), @() (cstrcat(result, "]")));
            jsonstringifyinnerresult = result;
            return;
        end
        if isstruct(anythinginner)
            if (numel(fieldnames(anythinginner)) == 0)
                jsonstringifyinnerresult = "{}";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("{", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("{"));
            objectkeys = fieldnames(anythinginner);
            for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                objectkey = objectkeys{objectentryindex};
                objectvalue = anythinginner.(objectkey);
                result = cstrcat(result, """", objectkey, """: ", jsonstringifyinner(objectvalue, prettyinner, indentinner));
                if (objectentryindex ~= numel(objectkeys))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "}")), @() (cstrcat(result, "}")));
            jsonstringifyinnerresult = result;
            return;
        end
        jsonstringifyinnerresult = "null";
    end
    prettydefault = false;
    indentdefault = "    ";
    pretty = prettydefault;
    indent = indentdefault;
    if isempty(varargin)
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    optionalargument = varargin{1};
    if (islogical(optionalargument) && (optionalargument == false))
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    if isstruct(optionalargument)
        pretty = optionalchaining(optionalargument, "pretty");
        indent = optionalchaining(optionalargument, "indent");
        pretty = ternary(isempty(pretty), @() (prettydefault), @() (pretty));
        indent = ternary(isempty(indent), @() (indentdefault), @() (indent));
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    if (islogical(optionalargument) && (optionalargument == true))
        if (numel(varargin) >= 2)
            optionalargument2 = varargin{2};
            if (ischar(optionalargument2) || ischar(optionalargument2))
                indent = optionalargument2
            end
        end
        pretty = optionalargument;
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
end

function sprint(varargin)
    currentresult = "";
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if (iscell(argument) && (numel(argument) == 1))
            argumentnew = jsonstringify(argument{1});
            argumentnew = strrep(argumentnew, """{", "{");
            argumentnew = strrep(argumentnew, """[", "[");
            argumentnew = strrep(argumentnew, "}""", "}");
            argumentnew = strrep(argumentnew, "]""", "]");
            currentresult = cstrcat(currentresult, argumentnew);
            continue;
        end
        if (~ischar(argument) && ~ischar(argument))
            error("Non string argument must be wrapped in {}");
            continue;
        end
        currentresult = cstrcat(currentresult, argument);
    end
    disp(currentresult);
end

function newallowedobjectkeygeneratorresult = newallowedobjectkeygenerator()
    iterator = 1;
    function generateallowedobjectkeyinnerresult = generateallowedobjectkeyinner(anything)
        if isnumeric(anything)
            anything = strrep(num2str(anything), ".", "point");
            anything = strrep(anything, "-", "negative");
            generateallowedobjectkeyinnerresult = cstrcat("key", anything);
            return;
        end
        newobjectkeystring = jsonstringify(anything);
        if ischar(anything) || ischar(anything)
            if regexp(anything, "^[0-9]*$")
                anything = strrep(num2str(anything), ".", "point");
                anything = strrep(anything, "-", "negative");
                generateallowedobjectkeyinnerresult = cstrcat("key", anything);
                return;
            end
        end
        allowedobjectkey = regexprep(newobjectkeystring, "^[^a-zA-Z]+|[^a-zA-Z0-9_]", "");
        if (allowedobjectkey ~= "")
            generateallowedobjectkeyinnerresult = allowedobjectkey;
            return;
        end
        generateallowedobjectkeyinnerresult = cstrcat("key", num2str(iterator));
        iterator = iterator + 1;
    end
    newallowedobjectkeygeneratorresult = @generateallowedobjectkeyinner;
end
generateallowedobjectkey = newallowedobjectkeygenerator();

function spreadobjectresult = spreadobject(varargin)
    newobject = struct();
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if isstruct(argument)
            objectkeys = fieldnames(argument);
            for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                objectkey = objectkeys{objectentryindex};
                objectvalue = argument.(objectkey);
                newobject.(objectkey) = objectvalue;
            end
            continue;
        end
        if iscell(argument)
            for arrayitemindex = (1:1:numel(argument)) % (start:step:stop)
                arrayitem = argument{arrayitemindex};
                newobject.(cstrcat("index", num2str(arrayitemindex))) = arrayitem;
            end
            continue;
        end
    end
    spreadobjectresult = newobject;
end

function arraymapv1result = arraymapv1(callbackfunction, anarray)
    % JavaScript-like Array.map() function
    newarray = {};
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        newarrayitem = callbackfunction(arrayitem, arrayitemindex, anarray);
        newarray{end + 1} = newarrayitem;
    end
    arraymapv1result = newarray;
end

function arraymapv2result = arraymapv2(callbackfunction, anarray)
    % JavaScript-like Array.map() function
    newarray = {};
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        newarray{end + 1} = callbackfunction(arrayitem, arrayitemindex, anarray);
    end
    arraymapv2result = newarray;
end

sprint(sprintf("\n"), "% JavaScript-like Array.map() in GNU Octave Cell-Array");

numbers = {12, 34, 27, 23, 65, 93, 36, 87, 4, 254};
sprint("numbers: ", jsonstringify(numbers));

sprint("% using JavaScript-like Array.map() function ""arraymapv1""");

numberslabeled = arraymapv1(@(number, varargin) ternary((mod(number, 2) == 0), @() (struct(generateallowedobjectkey(number), {"even"})), @() (struct(generateallowedobjectkey(number), {"odd"}))), numbers);
sprint("labeled numbers: ", jsonstringify(numberslabeled, struct("pretty", {true})));
% labeled numbers: [
%     {
%         "key12": "even"
%     },
%     {
%         "key34": "even"
%     },
%     {
%         "key27": "odd"
%     },
%     {
%         "key23": "odd"
%     },
%     {
%         "key65": "odd"
%     },
%     {
%         "key93": "odd"
%     },
%     {
%         "key36": "even"
%     },
%     {
%         "key87": "odd"
%     },
%     {
%         "key4": "even"
%     },
%     {
%         "key254": "even"
%     }
% ]

sprint("% using JavaScript-like Array.map() function ""arraymapv2""");

numberslabeled = arraymapv2(@(number, varargin) ternary((mod(number, 2) == 0), @() (struct(generateallowedobjectkey(number), {"even"})), @() (struct(generateallowedobjectkey(number), {"odd"}))), numbers);
sprint("labeled numbers: ", jsonstringify(numberslabeled, struct("pretty", {true})));
% labeled numbers: [
%     {
%         "key12": "even"
%     },
%     {
%         "key34": "even"
%     },
%     {
%         "key27": "odd"
%     },
%     {
%         "key23": "odd"
%     },
%     {
%         "key65": "odd"
%     },
%     {
%         "key93": "odd"
%     },
%     {
%         "key36": "even"
%     },
%     {
%         "key87": "odd"
%     },
%     {
%         "key4": "even"
%     },
%     {
%         "key254": "even"
%     }
% ]

sprint("% using GNU Octave Array.map() built-in function ""cellfun""");

numberslabeled = cellfun(@(number) ternary((mod(number, 2) == 0), @() (struct(generateallowedobjectkey(number), {"even"})), @() (struct(generateallowedobjectkey(number), {"odd"}))), numbers, "UniformOutput", false);
sprint("labeled numbers: ", jsonstringify(numberslabeled, struct("pretty", {true})));
% labeled numbers: [
%     {
%         "key12": "even"
%     },
%     {
%         "key34": "even"
%     },
%     {
%         "key27": "odd"
%     },
%     {
%         "key23": "odd"
%     },
%     {
%         "key65": "odd"
%     },
%     {
%         "key93": "odd"
%     },
%     {
%         "key36": "even"
%     },
%     {
%         "key87": "odd"
%     },
%     {
%         "key4": "even"
%     },
%     {
%         "key254": "even"
%     }
% ]

sprint(sprintf("\n"), "% JavaScript-like Array.map() in GNU Octave Cell-Array of structs");

products = { ...
    struct( ...
        "code", {"pasta"}, ...
        "price", {321} ...
    ), ...
    struct( ...
        "code", {"bubble_gum"}, ...
        "price", {233} ...
    ), ...
    struct( ...
        "code", {"potato_chips"}, ...
        "price", {5} ...
    ), ...
    struct( ...
        "code", {"towel"}, ...
        "price", {499} ...
    ) ...
};
sprint("products: ", jsonstringify(products, struct("pretty", {true})));

sprint("% using JavaScript-like Array.map() function ""arraymapv1""");

productslabeled = arraymapv1(@(product, varargin) spreadobject(product, struct("label", {ternary((product.("price") > 100), @() ("expensive"), @() ("cheap"))})), products);
sprint("labeled products: ", jsonstringify(productslabeled, struct("pretty", {true})));
% labeled products: [
%     {
%         "code": "pasta",
%         "price": 321,
%         "label": "expensive"
%     },
%     {
%         "code": "bubble_gum",
%         "price": 233,
%         "label": "expensive"
%     },
%     {
%         "code": "potato_chips",
%         "price": 5,
%         "label": "cheap"
%     },
%     {
%         "code": "towel",
%         "price": 499,
%         "label": "expensive"
%     }
% ]

sprint("% using JavaScript-like Array.map() function ""arraymapv2""");

productslabeled = arraymapv2(@(product, varargin) spreadobject(product, struct("label", {ternary((product.("price") > 100), @() ("expensive"), @() ("cheap"))})), products);
sprint("labeled products: ", jsonstringify(productslabeled, struct("pretty", {true})));
% labeled products: [
%     {
%         "code": "pasta",
%         "price": 321,
%         "label": "expensive"
%     },
%     {
%         "code": "bubble_gum",
%         "price": 233,
%         "label": "expensive"
%     },
%     {
%         "code": "potato_chips",
%         "price": 5,
%         "label": "cheap"
%     },
%     {
%         "code": "towel",
%         "price": 499,
%         "label": "expensive"
%     }
% ]

sprint("% using GNU Octave Array.map() built-in function ""cellfun""");

productslabeled = cellfun(@(product) spreadobject(product, struct("label", {ternary((product.("price") > 100), @() ("expensive"), @() ("cheap"))})), products, "UniformOutput", false);
sprint("labeled products: ", jsonstringify(productslabeled, struct("pretty", {true})));
% labeled products: [
%     {
%         "code": "pasta",
%         "price": 321,
%         "label": "expensive"
%     },
%     {
%         "code": "bubble_gum",
%         "price": 233,
%         "label": "expensive"
%     },
%     {
%         "code": "potato_chips",
%         "price": 5,
%         "label": "cheap"
%     },
%     {
%         "code": "towel",
%         "price": 499,
%         "label": "expensive"
%     }
% ]
